File: native-array-memory/js-2/stack.js
"use strict";

function Stack (size, init_value) {
  Buffer.call(this, size, init_value);
  this.cursor = 0;
};

Stack.prototype = Object.create(Buffer.prototype);

Stack.prototype.push = function (value) {
  this.set(this.cursor, value);
  this.add_cursor(1);
};

Stack.prototype.pop = function () {
  this.add_cursor(-1);
  return this.get(this.cursor);
};

File: native-array-memory/clojure/interpreter/src/interpreter/stack.clj
(ns interpreter.stack
  (:use interpreter.buffer))

(def make-stack make-buffer)

(defn stack-push [stack value]
  (buffer-set stack
              (buffer-get-cursor stack)
              value)
  (buffer-add-cursor stack 1))

(defn stack-pop [stack]
  (buffer-add-cursor stack -1)
  (buffer-get stack
              (buffer-get-cursor stack)))
